File: backend/src/services/authService.ts
import bcrypt from 'bcrypt';
import jwt from 'jsonwebtoken';
import { v4 as uuidv4 } from 'uuid';
import { query } from '../config/database';
import { config } from '../config';
import { User, UserRole, JWTPayload } from '../types';
import { AppError } from '../middleware/errorHandler';
import { logger } from '../utils/logger';

const SALT_ROUNDS = 10;

export interface RegisterData {
  username: string;
  email: string;
  password: string;
  role: UserRole;
  displayName?: string;
  sportCategories?: string[];
}

export interface LoginData {
  username: string;
  password: string;
}

export interface AuthTokens {
  accessToken: string;
  refreshToken: string;
  user: Omit<User, 'password_hash'>;
}

export class AuthService {
  async register(data: RegisterData): Promise<AuthTokens> {
    const { username, email, password, role, displayName, sportCategories } = data;

    // Check if user already exists
    const existingUser = await query(
      'SELECT id FROM users WHERE username = $1 OR email = $2',
      [username, email]
    );

    if (existingUser.rows.length > 0) {
      throw new AppError('Username or email already exists', 409);
    }

    // Hash password
    const passwordHash = await bcrypt.hash(password, SALT_ROUNDS);

    // Create user
    const userResult = await query(
      `INSERT INTO users (username, email, password_hash, role, display_name)
       VALUES ($1, $2, $3, $4, $5)
       RETURNING id, username, email, role, display_name, bio, avatar_url, banner_url,
                 created_at, updated_at, is_verified, is_banned`,
      [username, email, passwordHash, role, displayName || username]
    );

    const user = userResult.rows[0];

    // If athlete, create athlete profile
    if (role === UserRole.ATHLETE) {
      const streamKey = this.generateStreamKey();
      await query(
        `INSERT INTO athlete_profiles (user_id, sport_categories, stream_key, rtmp_url)
         VALUES ($1, $2, $3, $4)`,
        [user.id, sportCategories || [], streamKey, `${config.rtmp.serverUrl}/${streamKey}`]
      );
    }

    // Generate tokens
    const tokens = this.generateTokens({
      userId: user.id,
      username: user.username,
      role: user.role,
    });

    logger.info(`User registered: ${username} (${role})`);

    return {
      ...tokens,
      user,
    };
  }

  async login(data: LoginData): Promise<AuthTokens> {
    const { username, password } = data;

    // Find user
    const userResult = await query(
      `SELECT id, username, email, password_hash, role, display_name, bio, avatar_url,
              banner_url, created_at, updated_at, is_verified, is_banned
       FROM users WHERE username = $1`,
      [username]
    );

    if (userResult.rows.length === 0) {
      throw new AppError('Invalid username or password', 401);
    }

    const user = userResult.rows[0];

    // Check if banned
    if (user.is_banned) {
      throw new AppError('Account has been banned', 403);
    }

    // Verify password
    const isPasswordValid = await bcrypt.compare(password, user.password_hash);
    if (!isPasswordValid) {
      throw new AppError('Invalid username or password', 401);
    }

    // Generate tokens
    const tokens = this.generateTokens({
      userId: user.id,
      username: user.username,
      role: user.role,
    });

    // Remove password hash from user object
    const { password_hash, ...userWithoutPassword } = user;

    logger.info(`User logged in: ${username}`);

    return {
      ...tokens,
      user: userWithoutPassword,
    };
  }

  async refreshToken(refreshToken: string): Promise<{ accessToken: string }> {
    try {
      const decoded = jwt.verify(refreshToken, config.jwt.refreshSecret) as JWTPayload;

      // Verify user still exists
      const userResult = await query(
        'SELECT id, username, role, is_banned FROM users WHERE id = $1',
        [decoded.userId]
      );

      if (userResult.rows.length === 0 || userResult.rows[0].is_banned) {
        throw new AppError('Invalid refresh token', 401);
      }

      const user = userResult.rows[0];

      // Generate new access token
      const accessToken = jwt.sign(
        {
          userId: user.id,
          username: user.username,
          role: user.role,
        },
        config.jwt.secret,
        { expiresIn: config.jwt.expiry }
      );

      return { accessToken };
    } catch (error) {
      throw new AppError('Invalid refresh token', 401);
    }
  }

  async getMe(userId: string): Promise<any> {
    const userResult = await query(
      `SELECT u.id, u.username, u.email, u.role, u.display_name, u.bio, u.avatar_url,
              u.banner_url, u.created_at, u.updated_at, u.is_verified,
              ap.sport_categories, ap.stream_key, ap.rtmp_url, ap.follower_count,
              ap.total_view_count, ap.is_live
       FROM users u
       LEFT JOIN athlete_profiles ap ON u.id = ap.user_id
       WHERE u.id = $1`,
      [userId]
    );

    if (userResult.rows.length === 0) {
      throw new AppError('User not found', 404);
    }

    return userResult.rows[0];
  }

  private generateTokens(payload: JWTPayload): { accessToken: string; refreshToken: string } {
    const accessToken = jwt.sign(payload, config.jwt.secret, {
      expiresIn: config.jwt.expiry,
    });

    const refreshToken = jwt.sign(payload, config.jwt.refreshSecret, {
      expiresIn: config.jwt.refreshExpiry,
    });

    return { accessToken, refreshToken };
  }

  private generateStreamKey(): string {
    return `${uuidv4().replace(/-/g, '')}_${Date.now()}`;
  }
}

export const authService = new AuthService();

File: backend-go/internal/service/auth_service.go
package service

import (
	"errors"
	"fmt"

	"github.com/google/uuid"
	"github.com/phitonias/streak/internal/config"
	"github.com/phitonias/streak/internal/database"
	"github.com/phitonias/streak/internal/models"
	"github.com/phitonias/streak/internal/utils"
	"gorm.io/gorm"
)

type AuthService struct {
	cfg *config.Config
}

func NewAuthService(cfg *config.Config) *AuthService {
	return &AuthService{cfg: cfg}
}

type RegisterInput struct {
	Username        string
	Email           string
	Password        string
	Role            models.UserRole
	DisplayName     string
	SportCategories []string
}

type AuthResponse struct {
	AccessToken  string      `json:"accessToken"`
	RefreshToken string      `json:"refreshToken"`
	User         interface{} `json:"user"`
}

func (s *AuthService) Register(input RegisterInput) (*AuthResponse, error) {
	// Check if user already exists
	var existingUser models.User
	result := database.DB.Where("username = ? OR email = ?", input.Username, input.Email).First(&existingUser)
	if result.Error == nil {
		return nil, errors.New("username or email already exists")
	} else if !errors.Is(result.Error, gorm.ErrRecordNotFound) {
		return nil, fmt.Errorf("database error: %w", result.Error)
	}

	// Hash password
	hashedPassword, err := utils.HashPassword(input.Password)
	if err != nil {
		return nil, fmt.Errorf("failed to hash password: %w", err)
	}

	// Set display name if not provided
	displayName := input.DisplayName
	if displayName == "" {
		displayName = input.Username
	}

	// Create user
	user := models.User{
		Username:     input.Username,
		Email:        input.Email,
		PasswordHash: hashedPassword,
		Role:         input.Role,
		DisplayName:  &displayName,
	}

	if err := database.DB.Create(&user).Error; err != nil {
		return nil, fmt.Errorf("failed to create user: %w", err)
	}

	// If athlete, create athlete profile
	if input.Role == models.RoleAthlete {
		streamKey := generateStreamKey()
		rtmpURL := fmt.Sprintf("%s/%s", s.cfg.RTMP.ServerURL, streamKey)

		athleteProfile := models.AthleteProfile{
			UserID:          user.ID,
			SportCategories: input.SportCategories,
			StreamKey:       streamKey,
			RTMPURL:         &rtmpURL,
		}

		if err := database.DB.Create(&athleteProfile).Error; err != nil {
			return nil, fmt.Errorf("failed to create athlete profile: %w", err)
		}

		user.AthleteProfile = &athleteProfile
	}

	// Generate tokens
	tokens, err := utils.GenerateTokenPair(
		user.ID,
		user.Username,
		user.Role,
		s.cfg.JWT.Secret,
		s.cfg.JWT.RefreshSecret,
		s.cfg.JWT.Expiry,
		s.cfg.JWT.RefreshExpiry,
	)
	if err != nil {
		return nil, fmt.Errorf("failed to generate tokens: %w", err)
	}

	return &AuthResponse{
		AccessToken:  tokens.AccessToken,
		RefreshToken: tokens.RefreshToken,
		User:         sanitizeUser(&user),
	}, nil
}

func (s *AuthService) Login(username, password string) (*AuthResponse, error) {
	// Find user
	var user models.User
	result := database.DB.Preload("AthleteProfile").Where("username = ?", username).First(&user)
	if result.Error != nil {
		if errors.Is(result.Error, gorm.ErrRecordNotFound) {
			return nil, errors.New("invalid username or password")
		}
		return nil, fmt.Errorf("database error: %w", result.Error)
	}

	// Check if banned
	if user.IsBanned {
		return nil, errors.New("account has been banned")
	}

	// Verify password
	if !utils.CheckPassword(password, user.PasswordHash) {
		return nil, errors.New("invalid username or password")
	}

	// Generate tokens
	tokens, err := utils.GenerateTokenPair(
		user.ID,
		user.Username,
		user.Role,
		s.cfg.JWT.Secret,
		s.cfg.JWT.RefreshSecret,
		s.cfg.JWT.Expiry,
		s.cfg.JWT.RefreshExpiry,
	)
	if err != nil {
		return nil, fmt.Errorf("failed to generate tokens: %w", err)
	}

	return &AuthResponse{
		AccessToken:  tokens.AccessToken,
		RefreshToken: tokens.RefreshToken,
		User:         sanitizeUser(&user),
	}, nil
}

func (s *AuthService) RefreshToken(refreshToken string) (string, error) {
	// Validate refresh token
	claims, err := utils.ValidateToken(refreshToken, s.cfg.JWT.RefreshSecret)
	if err != nil {
		return "", errors.New("invalid refresh token")
	}

	// Verify user still exists and is not banned
	var user models.User
	if err := database.DB.Where("id = ? AND is_banned = false", claims.UserID).First(&user).Error; err != nil {
		return "", errors.New("invalid refresh token")
	}

	// Generate new access token
	tokens, err := utils.GenerateTokenPair(
		user.ID,
		user.Username,
		user.Role,
		s.cfg.JWT.Secret,
		s.cfg.JWT.RefreshSecret,
		s.cfg.JWT.Expiry,
		s.cfg.JWT.RefreshExpiry,
	)
	if err != nil {
		return "", fmt.Errorf("failed to generate token: %w", err)
	}

	return tokens.AccessToken, nil
}

func (s *AuthService) GetMe(userID uuid.UUID) (interface{}, error) {
	var user models.User
	result := database.DB.Preload("AthleteProfile").Where("id = ?", userID).First(&user)
	if result.Error != nil {
		return nil, fmt.Errorf("user not found: %w", result.Error)
	}

	return sanitizeUser(&user), nil
}

// Helper functions
func generateStreamKey() string {
	return fmt.Sprintf("%s_%d", uuid.New().String()[:8], uuid.New().ID())
}

func sanitizeUser(user *models.User) map[string]interface{} {
	result := map[string]interface{}{
		"id":           user.ID,
		"username":     user.Username,
		"email":        user.Email,
		"role":         user.Role,
		"display_name": user.DisplayName,
		"bio":          user.Bio,
		"avatar_url":   user.AvatarURL,
		"banner_url":   user.BannerURL,
		"created_at":   user.CreatedAt,
		"is_verified":  user.IsVerified,
	}

	if user.AthleteProfile != nil {
		result["sport_categories"] = user.AthleteProfile.SportCategories
		result["stream_key"] = user.AthleteProfile.StreamKey
		result["rtmp_url"] = user.AthleteProfile.RTMPURL
		result["follower_count"] = user.AthleteProfile.FollowerCount
		result["total_view_count"] = user.AthleteProfile.TotalViewCount
		result["is_live"] = user.AthleteProfile.IsLive
	}

	return result
}
